### ISSUE TITLE
phpMyAdmin SQL Dump - Export/Import (CREATE TABLE IF NOT EXISTS)

### ISSUE THREAD

**Issue by llcool** [phpmyadmin/phpmyadmin#11407]
-- phpMyAdmin SQL Dump
-- version 4.2.12deb2
-- http://www.phpmyadmin.net
--
-- Host: localhost
-- Generation Time: Aug 12, 2015 at 10:09 AM
-- Server version: 5.6.25-0ubuntu0.15.04.1
-- PHP Version: 5.6.4-4ubuntu6.2

CREATE TABLE IF NOT EXISTS `alert` (
  `alert_id` int(11) NOT NULL,
  `user_id` int(11) DEFAULT NULL,
  `email` varchar(80) DEFAULT NULL
) ENGINE=MyISAM AUTO_INCREMENT=29 DEFAULT CHARSET=latin1;
.
.
.
ALTER TABLE `alert`
 ADD PRIMARY KEY (`alert_id`);

If table structure exists ALTER TABLE will fail with duplicate key and the import exits any other modification will not run.

Why not go back to CREATE TABLE with PRIMARY KEY.

CREATE TABLE IF NOT EXISTS `alert` (
  `alert_id` int(11) NOT NULL,
  `user_id` int(11) DEFAULT NULL,
  `email` varchar(80) DEFAULT NULL,
  PRIMARY KEY (`alert_id`)
) ENGINE=MyISAM AUTO_INCREMENT=29 DEFAULT CHARSET=latin1;

**Comment by lem9** [comment]
There was a reason to split the statement like this.

**Comment by llcool** [comment]
lem9, You statement seems incomplete, or am I missing something...

**Comment by nisargjhaveri** [comment]
See, #6366 and #5703.

But, yes, it is really frustrating when importing a dump into an already existing databases and it fails with similar error. Can we do something about it?

**Comment by llcool** [comment]
Thank nisargjhaveri, for the info.

Why can't we have ADD PRIMARY KEY IF NOT EXIST

**Comment by lem9** [comment]
@llcool When I have time, I'll search to complete my above comment and clarify the reason why there are two statements.  About your suggestion, looking at http://dev.mysql.com/doc/refman/5.6/en/alter-table.html I don't see that it would be possible.

**Comment by lem9** [comment]
Thanks to @nisargjhaveri comment I have read in #6366 that "creating index and foreign keys only after inserting data makes inserting much efficient. "; but the case @llcool was not taken into account.

**Comment by nisargjhaveri** [comment]
Yes, so I can think of couple of ways in which we can probably improve this.

1. Actually check if the constraint or key exists or not, when doing alter table. (I don't know how good or bad it would be)
2. Give an option on export page to choose between whether to separate indexes and table structure or not. (Also quoting, https://github.com/phpmyadmin/phpmyadmin/issues/5703#issuecomment-124648245)
3. On a slightly different note, when exporting only structure of a table, we don't actually have to separate index creation and rest of table structure, right?

**Comment by lem9** [comment]
@nisargjhaveri About your 1), it depends: do you want this export to be readable from tools other that phpMyAdmin? If I understand correctly your suggestion, it would be a change of logic at import time from phpMyAdmin. About your 2), it would be a good idea and the option could be "Do you plan to import this into an existing database, where tables already exist".  About your 3), the new option suggested in 2) could be shown only when exporting data.  Do you wish to work on this?

**Comment by nisargjhaveri** [comment]
1) No, I was not talking about change in import logic, I was suggesting something like [this question on stackowerflow](http://stackoverflow.com/questions/16792516/if-foreign-key-not-exist-then-add-foreign-key-constraintor-drop-a-foreign-key-c).

2) If option is only what you suggested, that can be done by just exporting data. I was thinking about a fallback to standard export where we don't do any alter operations after data for efficiency, and keys are defined in table creating statement itself. Sorry if I misunderstood your comment on this.

3) I was thinking to set the behaviour in 2) to be default when exporting only structure. No extra options.

Yes, I can work on this.

**Comment by lem9** [comment]
1) I don't believe that this syntax exists in MySQL (starting a statement with IF NOT EXISTS).  2) Yes there would be an option producing the key definitions in the CREATE TABLE statement, but I'm not sure that this should be the default, as it's not efficient.

**Comment by lem9** [comment]
Found this: http://dev.mysql.com/doc/refman/5.6/en/if.html but I believe it's only applicable inside routines.

**Comment by nisargjhaveri** [comment]
Yes @lem9, looks like the `IF` syntax is indeed applicable only inside routines.

Also, what does `IGNORE` option do in `ALTER TABLE`? https://dev.mysql.com/doc/refman/5.5/en/alter-table.html#idm140186428351664

Isn't that we want or is it only for existing conflicting rows and not actual alter table operation.

**Comment by lem9** [comment]
The IGNORE option in ALTER TABLE looks promising, please test it.

**Comment by nisargjhaveri** [comment]
Sure. I'll test it in a while.

**Comment by nisargjhaveri** [comment]
`IGNORE` doesn't work, it still gives "Multiple primary key defined" error.

Another possible workaround might be to create a temporary procedure, call it from script itself and then drop it, as someone suggested [here](https://bugs.mysql.com/bug.php?id=15037).

The single create table statement looks more human comprehensible, and friendly to some database comparison tools, I don't think that is the solution to the problem reported. It would be nice to have this, though.
But I think, if possible, there at least should be no error if I run an exported script, especially when the table creation already have `IF NOT EXISTS`.

**Comment by lem9** [comment]
So, offering an option at export time to generate a complete CREATE TABLE statement with indexes (with a warning about reduced efficiency) is the best idea so far.

**Comment by madhuracj** [comment]
I don't know whether a separate option is required. Generally, the use of `IF NOT EXISTS` is a good indication of whether the user wants to import the SQL file into a database whether tables already exist.

**Comment by nisargjhaveri** [comment]
I share @madhuracj's view on this.

**Comment by nisargjhaveri** [comment]
See #11419.

**Comment by lem9** [comment]
@madhuracj But in our export dialog, the IF NOT EXISTS clause of "Add CREATE TABLE statement" is ticked by default, so most exports (at least those done in Quick mode) are likely to generate this clause. Or am I missing something?

**Comment by nisargjhaveri** [comment]
Yes, but we can probably make it unchecked by default, and also add efficiency warning.
The main reason to do this is, even if IF NOT EXISTS clause is there, it is useless, because keys are too common, and the import will fail there. So, if it is anyway going to fail, so why not on table creation. And if anyone wants IF NOT EXISTS specifically, they can enable it.

**Comment by madhuracj** [comment]
I also think it make sense to change the default of  `IF NOT EXISTS`

**Comment by lem9** [comment]
Ok. Nisarg can you update your commit as in your suggestion?

**Comment by nisargjhaveri** [comment]
Okay. I'll make it unchecked by default.
Also, we want to add efficiency warning along with IF NOT EXISTS, right?

**Comment by lem9** [comment]
Yes, an efficiency warning next to IF NOT EXISTS is important.

**Comment by lem9** [comment]
By the way, these changes would go to 4.5, right?

**Comment by nisargjhaveri** [comment]
Yes, these changes are for 4.5.
